pkg/kn/commands/service/import.go

```diff
@@ -0,0 +1,143 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/util/retry"
+
+	clientv1alpha1 "knative.dev/client/pkg/apis/client/v1alpha1"
+	"knative.dev/client/pkg/kn/commands"
+	clientservingv1 "knative.dev/client/pkg/serving/v1"
+	"knative.dev/pkg/kmeta"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// NewServiceImportCommand returns a new command for importing a service.
+func NewServiceImportCommand(p *commands.KnParams) *cobra.Command {
+	var waitFlags commands.WaitFlags
+
+	command := &cobra.Command{
+		Use:   "import FILENAME",
+		Short: "Import a service and its revisions",
+		Example: `
+ # Import a service from YAML file
+ kn service import /path/to/file.yaml
+
+ # Import a service from JSON file
+ kn service import /path/to/file.json`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return errors.New("'kn service import' requires filename of import file as single argument")
+			}
+			filename := args[0]
+
+			namespace, err := p.GetNamespace(cmd)
+			if err != nil {
+				return err
+			}
+
+			client, err := p.NewServingClient(namespace)
+			if err != nil {
+				return err
+			}
+
+			return importWithOwnerRef(client, filename, cmd.OutOrStdout(), waitFlags)
+		},
+	}
+	flags := command.Flags()
+	commands.AddNamespaceFlags(flags, false)
+	waitFlags.AddConditionWaitFlags(command, commands.WaitDefaultTimeout, "import", "service", "ready")
+
+	return command
+}
+
+func importWithOwnerRef(client clientservingv1.KnServingClient, filename string, out io.Writer, waitFlags commands.WaitFlags) error {
+	var export clientv1alpha1.Export
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 512)
+	err = decoder.Decode(&export)
+	if err != nil {
+		return err
+	}
+	if export.Spec.Service.Name == "" {
+		return fmt.Errorf("provided import file doesn't contain service name, please note that only kn' custom Export format is supported")
+	}
+
+	serviceName := export.Spec.Service.Name
+
+	// Return error if service already exists
+	svcExists, err := serviceExists(client, serviceName)
+	if err != nil {
+		return err
+	}
+	if svcExists {
+		return fmt.Errorf("cannot import service '%s' in namespace '%s' because the service already exists",
+			serviceName, client.Namespace())
+	}
+
+	err = client.CreateService(&export.Spec.Service)
```

Comment:
when we create a service like this, then an initial revision is created. This is fine, but I think we might get into ordering issues as the generation of this revision would be 1 (and which might conflict with the revision history that we are importing, where there would be also 1 generation). At least this is what I suspect. (would need to be verified)

Reply:
Well, your suspicion is exactly on point here, revision history is mangled after import. I'll try to double-check if there's an alternative to the current import order. 
```
➜  client git:(service-import) kn revision list                 
NAME            SERVICE   TRAFFIC   TAGS   GENERATION   AGE   CONDITIONS   READY   REASON
hello-xbgkn-2   hello     25%              2            60s   4 OK / 4     True    
hello-brqcf-3   hello     50%              1            60s   4 OK / 4     True    
hello-kphzr-1   hello     25%              1            60s   4 OK / 4     True   
```

Reply:
Currently revision list is sorted by 
```
// revisionListSortFunc sorts by namespace, service,  generation and name
```
In fact it's Configuration's generation, [here](// revisionListSortFunc sorts by namespace, service,  generation and name). That's in line with revision history in my comment above. However, I don't see any clean way how to reproduce the history in the import wihout leaning towards "replay" approach actually.